xaitk_saliency/impls/perturb_image/occlusion.py

```diff
@@ -0,0 +1,64 @@
+from typing import Dict, Any, Iterable, Tuple, List
+
+import PIL.Image
+import numpy as np
+
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+from xaitk_saliency.utils.masking import generate_block_masks
+
+
+class OcclusionBasedPerturb (PerturbImage):
+    """
+    Produce image perturbations based on hard, block-y occlusion areas as
+    generated by sliding a window of a configured size.
+
+    For multi-channel input, all channels are occluded at the same time or not
+    at all.
+    For example, if the input image is an RGB image, for regions masked, R, G
+    and B channels will be masked at the same time.
+    If the image was a 2x2 white image, one mask of just the upper-left pixel
+    might be as follows::
+
+        img = (
+            [[[255, 255, 255], [255, 255, 255]],
+             [[255, 255, 255], [255, 255, 255]]]
+         )
+        img_mask_example = (
+            [[[  0,   0,   0], [255, 255, 255]],
+             [[255, 255, 255], [255, 255, 255]]]
+        )
+
+    :param window_size: The block window size.
+    :param stride: The sliding window striding step.
+    """
+
+    def __init__(self, window_size: int = 50, stride: int = 20):
+        self.window_size = int(window_size)
+        self.stride = int(stride)
+
+    def perturb(
+        self,
+        ref_image: PIL.Image.Image
+    ) -> Tuple[List[PIL.Image.Image], np.ndarray]:
+        ref_mat = np.asarray(ref_image)
+        masks = mask_lhs = generate_block_masks(
```

Comment:
I realize that this is a change from the base implementation that moves the mask generation into the runtime method, however the previous implementation had a requirement that all input imagery match the generated mask size. This is a bit fragile as data in the wild is often of different sizes.

We can do something where we use result caching (`functools.lrucache`) to optimize this for repeat image size invocations.

Alternatively, is there instead a desire for this impl to generate a static mask grid and then "scale" that to the input imagery? This impl will create blocks of constant N-pixel squares. The percent impact on input imagery will however depend on the scale of the input. E.g. an image scaled to NxM will be masked differently than if it was scaled to (5*N)x(5*M).

Reply:
My understanding is that mask generation is somewhat of a bottleneck (i.e. we currently generate ~2K masks using nested for loops). I don't know what kind of performance hit we should expect if we have to regenerate the masks for each input image (although caching would probably help). Another hypothetical scenario would be for extremely high-resolution imagery- it might not be feasible to fit the full-resolution masks in memory. For our use case (which I think is pretty standard), input images were generally assumed to be re-sized to a standard size (e.g. 224x224 pixels).

I didn't quite understand the resizing argument. Either way, aren't we either resizing the masks or resizing the image, and I imagine there is potentially some equivalence there? In the case of re-sizing the mask, would we use some type of nearest neighbors interpolation? Other linear interpolation schemes would make the masks non-binary.

Reply:
I assume we don't want to resize the image since the current API states the output perturbations are the same resolution as the input. The argument there was so that we don't have implementations outputting potentially different image perturbations such that their scales are different than inputs for next-step consistency.

Let me try to add a more reasonable scale unit test for performance evaluation.

Reply:
@brianhhu Wanted to circle back on this after our discussion about investigating a streaming API going forward. Is this OK for this initial PR?

Reply:
Yes, I think this is fine for now- thanks.

Reply:
I think this is fine for now- thanks.